### ISSUE TITLE
How to log *all* messages

### ISSUE THREAD

**Issue by GavanWilhite** [vercel/ai#563]
I'm having a hard time finding a hook or way to log *all* message content including the latest completed message, user messages, function call responses, system messages, etc.

On the client it seems to be missing system messages and function call responses, on the server it seems to lose certain messages from createFunctionCallMessages.

Thoughts?

**Comment by 0x-Legend** [comment]
When you say log, do you mean just console.log, or do you mean upserting to database?

This is my process for upsetting *all* messages to my database:
```typescript
export function upsertChatsOnFinish({ isLoading, messages, chatId, userId, supabase }: upsertChatsOnFinishProps) {
  useUpdateEffect(() => {
    if (!isLoading) {
      let title: string;
      try { title = messages[0].content.substring(0, 100) } catch { return }  // needed when strict mode ON

      const id = chatId
      const createdAt = Date.now()
      const path = `/chat/${id}`  // TODO: maybe this should not begin with forward slash?
      const payload = { id, path, title, userId, messages, createdAt }

      supabase
        .from('chats')
        .upsert({ id: id, payload: payload as {} })
        .throwOnError()
        .then()
    }
  }, [isLoading]);
}
```

Example log:
```json
{
  "id": "HKENFWI",
  "path": "/chat/HKENFWI",
  "title": "Hello",
  "userId": "2354c37a-b7ef-441a-8440-935966f468d4",
  "messages": [
    {
      "id": "HKENFWI",
      "role": "user",
      "content": "Hello"
    },
    {
      "id": "ycK4Bep",
      "role": "assistant",
      "content": "Hi there! How can I assist you today?",
      "createdAt": "2023-10-22T09:30:55.217Z"
    },
    {
      "id": "HKENFWI",
      "role": "user",
      "content": "What is my win rate?"
    },
    {
      "id": "3HCYWSk",
      "role": "assistant",
      "content": "",
      "createdAt": "2023-10-22T09:31:03.396Z",
      "extractedCode": "const totalTrades = trades.length;\nconst winningTrades = trades.filter(trade => trade.realizedPnl_USD > 0).length;\nconst winRate = (winningTrades / totalTrades) * 100;\nconst result = `Your win rate is ${winRate.toFixed(2)}%`;\nresult;",
      "function_call": {
        "name": "analyze_trades",
        "arguments": ""
      },
      "extractedMethod": "JSON parsed successfully"
    },
    {
      "id": "LJItcua",
      "name": "analyze_trades",
      "role": "function",
      "content": "Your win rate is 30.62%"
    },
    {
      "id": "xIo0niZ",
      "role": "assistant",
      "content": "Your win rate is 30.62%.",
      "createdAt": "2023-10-22T09:31:09.682Z"
    },
    {
      "id": "HKENFWI",
      "role": "user",
      "content": "Thanks"
    },
    {
      "id": "rFvRaV5",
      "role": "assistant",
      "content": "You're welcome! If you have any more questions, feel free to ask.",
      "createdAt": "2023-10-22T09:31:13.976Z"
    }
  ],
  "createdAt": 1697967075052
}
```
It even has additional fields that I added the messages object like extractedMethod.

Note: I have removed the function call argument json.


TLDR: You can do useEffect on isLoading changing from true to false to log the messages.

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **None**  
Confidence: **High**

### Rationale
This is primarily a usage/support question about message logging, with a community-provided solution. It does not clearly request a product or documentation change or establish a concrete SDK defect.

### Evidence
- The reporter asks how to log all message types and requests guidance rather than identifying documented behavior that is violated.
- The only comment provides a client-side `useEffect` approach for persisting messages when loading finishes.
